modules/internal/ChapelArray.chpl

```diff
@@ -2037,7 +2061,18 @@ module ChapelArray {
             hi = this.domain.high+1;
       const newDom = {lo..hi};
       on this._value {
-        this._value.dsiReallocate(newDom);
+        if !this._value.dataAllocRange.member(hi) {
+          /* The new index is not in the allocated space.  We'll need to
+             realloc it. */
+          if this._value.dataAllocRange.length < this.domain.numIndices {
```

Comment:
If the array was being used as a non-vector-style array, the dataAllocRange can be out of date and smaller than this.domain.  In that case, it needs to be reset here.